sklearn/cluster/_hierarchical_fast.pyx

```diff
@@ -446,3 +449,98 @@ def single_linkage_label(L):
         raise ValueError("Input MST array must be sorted by weight")
 
     return _single_linkage_label(L)
+  
+
+# Implements MST-LINKAGE-CORE from https://arxiv.org/abs/1109.2378
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+def mst_linkage_core(
+        DTYPE_t [:, ::1] raw_data,
+        DistanceMetric dist_metric):
+    """
+    Compute the necessary elements of a minimum spanning
+    tree for computation of single linkage clustering. This
+    represents the MST-LINKAGE-CORE algorithm (Figure 6) from
+    *Modern hierarchical, agglomerative clustering algorithms*
+    by Daniel Mullner (https://arxiv.org/abs/1109.2378).
+
+    In contrast to the scipy implementation is never computes
+    a full distance matrix, generating distances only as they
+    are needed and releasing them when no longer needed.
+
+    Parameters
+    ----------
+    raw_data: array of shape (n_samples, n_features)
+        The array of feature data to be clustered. Must be C-aligned
+
+    dist_metric: DistanceMetric
+        A distance metric function (from sklearn.neighbors.dist_metrics)
+        used to compute distances.
+
+    Returns
+    -------
+    mst_core_data: array of shape (n_samples, 3)
+        An array providing information from which one
+        can either compute an MST, or the linkage hierarchy
+        very efficiently. See https://arxiv.org/abs/1109.2378
+        algorithm MST-LINKAGE-CORE for more details.
+    """
+    cdef:
+        np.int8_t[:] in_tree = np.zeros(dim, dtype=np.int8)
+        DTYPE_t * raw_data_ptr = (<DTYPE_t *> &raw_data[0, 0])
+        DTYPE_t[:, ::1] result = np.zeros((dim - 1, 3))
+
+        np.ndarray label_filter
+
+        ITYPE_t current_node = 0
+        ITYPE_t new_node
+        ITYPE_t i
+        ITYPE_t j
+        ITYPE_t dim = raw_data.shape[0]
```

Comment:
sorry about the painful merged (as you saw we renamed most files and changed imports).  The tests pass on my machine now, you just need to declare dim before it's used

Reply:
Ah, excellent; thanks. It's hard when everything starts breaking.